src/docker/assets.ts

```diff
@@ -221,16 +223,35 @@ provision:
     EOF
     fi
     export DEBIAN_FRONTEND=noninteractive
-    curl -fsSL https://get.docker.com | sh -s -- --channel {{dockerBinChannel}} --version {{dockerBinVersion}}
+    if [ {{srcType}} == "archive" ]; then
+      curl -fsSL https://get.docker.com | sh -s -- --channel {{srcArchiveChannel}} --version {{srcArchiveVersion}}
+    elif [ {{srcType}} == "image" ]; then
+      wget https://raw.githubusercontent.com/moby/moby/{{srcImageTag}}/contrib/init/systemd/docker.service \
+        https://raw.githubusercontent.com/moby/moby/v{{srcImageTag}}/contrib/init/systemd/docker.service \
+        -O /etc/systemd/system/docker.service || true
+      wget https://raw.githubusercontent.com/moby/moby/{{srcImageTag}}/contrib/init/systemd/docker.socket \
+        https://raw.githubusercontent.com/moby/moby/v{{srcImageTag}}/contrib/init/systemd/docker.socket \
+        -O /etc/systemd/system/docker.socket || true
+
+      mkdir -p /usr/local/bin
+      cp /tool/* /usr/local/bin/
+      sed -i 's|^ExecStart=.*|ExecStart=/usr/local/bin/dockerd -H fd://|' /etc/systemd/system/docker.service
+      sed -i 's|containerd.service||' /etc/systemd/system/docker.service
+      if ! getent group docker; then
+        groupadd --system docker
+      fi
+      systemctl daemon-reload
+      if ! systemctl enable --now docker; then
+        systemctl status docker.socket
+        systemctl status docker.service
+      fi
```

Comment:
Maybe instead of doing this we could just run `curl -fsSL https://get.docker.com | sh` and override binaries with ones from tooldir and restart daemon?

Reply:
I was considering that, but that could be problematic in cases where there are packaging changes between the version installed by `curl get.docker.com | sh` and the actual target version.